### DISCUSSION TITLE
How to use grad-cam on audio classify

### DISCUSSION THREAD

**Discussion by ThinkInFuture** [jacobgil/pytorch-grad-cam#311] [Q&A]
Hi 
from the project i get grad-cam is good for cv  on Advanced AI explainability for PyTorch
i am wondering if this can be worked on audio classify , also i am trying on this 
on audio part , also use CNN/ViT , usually first do MFCC on audio then send to Model
But in audio part , dont use cv2 for image read or write , instead is librosa
I not sure whether  grad-cam can be used for audio classify explainability , hope any suggestion can given to me

**Comment by jacobgil** [comment]
Hi,
What does the model look like ?
What's the input and the output tensors in the model ?

**Reply by ThinkInFuture** [reply]
the model structs as below
[model_1.txt](https://github.com/jacobgil/pytorch-grad-cam/files/9351259/model_1.txt)
the input is a song wav file take feature with fbank like below, specgram is the input

import torch
import torchaudio
import matplotlib.pyplot as plt
filename = "10001588.wav"
waveform,sample_rate = torchaudio.load(filename)
fbank = torchaudio.compliance.kaldi.fbank(waveform, htk_compat=True, sample_frequency=16000, use_energy=False,window_type='hanning', num_mel_bins=128, use_log_fbank=True,high_freq=8000.0, low_freq=20.0, dither=0.0, frame_length=32,frame_shift=16)
n_fft = 400.0
frame_length = n_fft / sample_rate * 1000.0
frame_shift = frame_length / 2.0
 
params = {
    "channel": 0,
    "dither": 0.0,
    "window_type": "hanning",
    "frame_length": frame_length,
    "frame_shift": frame_shift,
    "remove_dc_offset": False,
    "round_to_power_of_two": False,
    "sample_frequency": sample_rate,
}
 
specgram = torchaudio.compliance.kaldi.spectrogram(waveform, **params)

**Reply by ThinkInFuture** [reply]
specgram can be show to image with code
print("Shape of spectrogram: {}".format(specgram.size()))
fig, ax = plt.subplots(figsize=(12, 3))
im = ax.imshow(specgram, aspect="auto", origin="lower",interpolation='none')
plt.colorbar(im, ax=ax)
plt.xlabel("Frames")
plt.ylabel("Channels")
plt.tight_layout()

**Reply by ThinkInFuture** [reply]
I want to use grad-cam to do an interpretable application on audio classification. For audio classification, I usually do not send audio to the classification network first, but extract the Mel spectrum feature, which is called fbank. (Spectrum, which can be drawn by librosa) I want to use grad-cam to label the features used in the classification network's spectrogram through heatmaps to help me analyze the network.

**Reply by ThinkInFuture** [reply]
I also wonder if other teams have tried to apply this to audio classification before.

**Reply by jacobgil** [reply]
Thanks. 
How does the classification network look like?
Is it just a fully connected network?
What is its structure ?

**Reply by jacobgil** [reply]
does it use convolutions ?
